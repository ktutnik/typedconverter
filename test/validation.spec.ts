import reflect, { decorateMethod } from "tinspector"

import {
    convert,
    OptionalValidator,
    PartialValidator,
    requiredValidationVisitor,
    val,
    validate,
    ValidatorDecorator,
    validatorVisitor,
    VisitorInvocation,
    Result,
} from "../src"


describe("Optional & Partial Validation", () => {
    describe("Decorators", () => {
        it("Should able to decorate optional validator", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    @val.optional()
                    public birthday: Date
                ) { }
            }
            expect(reflect(AnimalClass)).toMatchSnapshot()
        })

        it("Should able to decorate partial validator", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }

            class AnimalController {
                @decorateMethod({})
                save(@val.partial(AnimalClass) data: AnimalClass) { }
            }
            expect(reflect(AnimalController)).toMatchSnapshot()
        })
    })

    describe("Primitive Type Validation", () => {

        it("Should validate undefined", () => {
            const result = convert(undefined, {
                path: "data",
                type: Number,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate null", () => {
            const result = convert(null, {
                path: "data",
                type: Number,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should valid if optional", () => {
            const result = convert(null, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: OptionalValidator }],
                path: "data",
                type: Number,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Object Validator", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                @val.optional()
                public birthday: Date
            ) { }
        }

        const option = { type: AnimalClass, visitors: [requiredValidationVisitor] }

        it("Should validate undefined property", () => {
            const result = convert({ id: undefined, name: "Mimi", deceased: "ON", birthday: "2018-1-1" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property", () => {
            const result = convert({ id: null, name: "Mimi", deceased: "ON", birthday: "2018-1-1" }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided null optional", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: null }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided undefined optional", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: undefined }, { ...option })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Array of Primitive Type Validation", () => {

        it("Should validate undefined", () => {
            const result = convert(["1", undefined, "3"], {
                path: "data",
                type: [Number],
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should validate null", () => {
            const result = convert(["1", null, "3"], {
                path: "data",
                type: [Number],
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
        it("Should valid if optional", () => {
            const result = convert(["1", null, "3"], {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: OptionalValidator }],
                path: "data",
                type: [Number],
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Array of Object Validator", () => {
        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                @val.optional()
                public birthday: Date
            ) { }
        }

        const option = { type: [AnimalClass], visitors: [requiredValidationVisitor] }

        it("Should validate undefined property", () => {
            const result = convert([undefined, { id: undefined, name: "Mimi", deceased: "ON", birthday: "2018-1-1" }], { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property", () => {
            const result = convert([undefined, { id: null, name: "Mimi", deceased: "ON", birthday: "2018-1-1" }], { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if optional", () => {
            const result = convert([undefined, { id: "123", name: "Mimi", deceased: "ON", birthday: null }], { ...option, decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: OptionalValidator }] })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Nested Object Validator", () => {
        @reflect.parameterProperties()
        class TagClass {
            constructor(
                public id: number,
                @val.optional()
                public name: string
            ) { }
        }

        @reflect.parameterProperties()
        class AnimalClass {
            constructor(
                public id: number,
                public name: string,
                public deceased: boolean,
                public birthday: Date,
                public tag: TagClass
            ) { }
        }

        const option = { type: AnimalClass, visitors: [requiredValidationVisitor] }

        it("Should validate undefined property", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-1-1", tag: { id: undefined, name: "The Tag" } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should validate null property", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-1-1", tag: { id: null, name: "The Tag" } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided null optional", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-1-1", tag: { id: "123", name: null } }, { ...option })
            expect(result).toMatchSnapshot()
        })

        it("Should valid if provided undefined optional", () => {
            const result = convert({ id: "123", name: "Mimi", deceased: "ON", birthday: "2018-1-1", tag: { id: "123", name: undefined } }, { ...option })
            expect(result).toMatchSnapshot()
        })
    })

    describe("Partial Validation", () => {
        it("Should pass empty field if parent marked with partial", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }
            const result = convert({ id: "123" }, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: PartialValidator }],
                path: "data",
                type: AnimalClass,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })

        it("Should skip optional validator on partial parent", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    @val.optional()
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }
            const result = convert({ id: "123" }, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: PartialValidator }],
                path: "data",
                type: AnimalClass,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })

        it("Should validate whole object even its partial and provided undefined", () => {
            @reflect.parameterProperties()
            class AnimalClass {
                constructor(
                    public id: number,
                    public name: string,
                    public deceased: boolean,
                    public birthday: Date
                ) { }
            }
            const result = convert(undefined, {
                decorators: [<ValidatorDecorator>{ type: "tc:validator", validator: PartialValidator }],
                path: "data",
                type: AnimalClass,
                visitors: [requiredValidationVisitor]
            })
            expect(result).toMatchSnapshot()
        })
    })

})

describe("Validator", () => {
    it("Should validate properly", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem.ipsum@gmail.com" }, { type: Dummy, })
        expect(result).toMatchObject({ value: { property: "lorem.ipsum@gmail.com" } })
    })

    it("Should able combine optional and email", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.optional()
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: undefined }, { type: Dummy, })
        expect(result).toMatchObject({ value: {} })
    })

    it("Should able combine optional and invalid email", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.optional()
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem ipsum" }, { type: Dummy, })
        expect(result).toMatchObject({ issues: [{ messages: ["Invalid email address"], path: "property" }] })
    })

    it("Should able combine optional and valid email", () => {
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.optional()
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem.ipsum@gmail.com" }, { type: Dummy, })
        expect(result).toMatchObject({ value: { property: "lorem.ipsum@gmail.com" } })
    })

    it("Should able to add custom visitor", () => {
        const visitor = (i:VisitorInvocation) => {
            if(i.type === String) return Result.create("lorem ipsum")
            else return i.proceed()
        }
        @reflect.parameterProperties()
        class Dummy {
            constructor(
                @val.optional()
                @val.email()
                public property: string
            ) { }
        }
        const result = validate({ property: "lorem.ipsum@gmail.com" }, { visitors: [visitor], type: Dummy, })
        expect(result).toMatchObject({ value: { property: "lorem ipsum" } })
    })
})
