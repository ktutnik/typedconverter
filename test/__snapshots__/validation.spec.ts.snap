// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optional & Partial Validation Array of Object Validator Should valid if optional 1`] = `
Object {
  "issues": undefined,
  "value": Array [
    undefined,
    AnimalClass {
      "deceased": true,
      "id": 123,
      "name": "Mimi",
    },
  ],
}
`;

exports[`Optional & Partial Validation Array of Object Validator Should validate null property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "0",
    },
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "1.id",
    },
  ],
  "value": Array [
    undefined,
    AnimalClass {
      "birthday": 2018-02-02,
      "deceased": true,
      "name": "Mimi",
    },
  ],
}
`;

exports[`Optional & Partial Validation Array of Object Validator Should validate undefined property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "0",
    },
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "1.id",
    },
  ],
  "value": Array [
    undefined,
    AnimalClass {
      "birthday": 2018-02-02,
      "deceased": true,
      "name": "Mimi",
    },
  ],
}
`;

exports[`Optional & Partial Validation Array of Primitive Type Validation Should valid if optional 1`] = `
Object {
  "issues": undefined,
  "value": Array [
    1,
    null,
    3,
  ],
}
`;

exports[`Optional & Partial Validation Array of Primitive Type Validation Should validate null 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data.1",
    },
  ],
  "value": Array [
    1,
    null,
    3,
  ],
}
`;

exports[`Optional & Partial Validation Array of Primitive Type Validation Should validate undefined 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data.1",
    },
  ],
  "value": Array [
    1,
    undefined,
    3,
  ],
}
`;

exports[`Optional & Partial Validation Decorators Should able to decorate optional validator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [
      Object {
        "decorators": Array [],
        "kind": "Parameter",
        "name": "id",
        "type": Number,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "kind": "Parameter",
        "name": "name",
        "type": String,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [],
        "kind": "Parameter",
        "name": "deceased",
        "type": Boolean,
        "typeClassification": "Primitive",
      },
      Object {
        "decorators": Array [
          Object {
            "type": "tc:validator",
            "validator": Symbol(tc:optional),
          },
        ],
        "kind": "Parameter",
        "name": "birthday",
        "type": Date,
        "typeClassification": "Primitive",
      },
    ],
  },
  "decorators": Array [
    Object {
      "type": "ParameterProperties",
    },
  ],
  "kind": "Class",
  "methods": Array [],
  "name": "AnimalClass",
  "properties": Array [
    Object {
      "decorators": Array [],
      "get": undefined,
      "kind": "Property",
      "name": "id",
      "set": undefined,
      "type": Number,
    },
    Object {
      "decorators": Array [],
      "get": undefined,
      "kind": "Property",
      "name": "name",
      "set": undefined,
      "type": String,
    },
    Object {
      "decorators": Array [],
      "get": undefined,
      "kind": "Property",
      "name": "deceased",
      "set": undefined,
      "type": Boolean,
    },
    Object {
      "decorators": Array [
        Object {
          "type": "tc:validator",
          "validator": Symbol(tc:optional),
        },
      ],
      "get": undefined,
      "kind": "Property",
      "name": "birthday",
      "set": undefined,
      "type": Date,
    },
  ],
  "type": AnimalClass,
  "typeClassification": "Class",
}
`;

exports[`Optional & Partial Validation Decorators Should able to decorate partial validator 1`] = `
Object {
  "ctor": Object {
    "kind": "Constructor",
    "name": "constructor",
    "parameters": Array [],
  },
  "decorators": Array [],
  "kind": "Class",
  "methods": Array [
    Object {
      "decorators": Array [
        Object {},
      ],
      "kind": "Method",
      "name": "save",
      "parameters": Array [
        Object {
          "decorators": Array [
            Object {
              "info": "Partial",
              "kind": "Override",
              "type": AnimalClass,
            },
            Object {
              "type": "tc:validator",
              "validator": Symbol(tc:partial),
            },
          ],
          "kind": "Parameter",
          "name": "data",
          "type": AnimalClass,
          "typeClassification": "Class",
        },
      ],
      "returnType": undefined,
      "typeClassification": undefined,
    },
  ],
  "name": "AnimalController",
  "properties": Array [],
  "type": AnimalController,
  "typeClassification": "Class",
}
`;

exports[`Optional & Partial Validation Nested Object Validator Should valid if provided null optional 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": 123,
    },
  },
}
`;

exports[`Optional & Partial Validation Nested Object Validator Should valid if provided undefined optional 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "id": 123,
    },
  },
}
`;

exports[`Optional & Partial Validation Nested Object Validator Should validate null property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "tag.id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "name": "The Tag",
    },
  },
}
`;

exports[`Optional & Partial Validation Nested Object Validator Should validate undefined property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "tag.id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "id": 123,
    "name": "Mimi",
    "tag": TagClass {
      "name": "The Tag",
    },
  },
}
`;

exports[`Optional & Partial Validation Object Validator Should valid if provided null optional 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "deceased": true,
    "id": 123,
    "name": "Mimi",
  },
}
`;

exports[`Optional & Partial Validation Object Validator Should valid if provided undefined optional 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "deceased": true,
    "id": 123,
    "name": "Mimi",
  },
}
`;

exports[`Optional & Partial Validation Object Validator Should validate empty string property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "name",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "id": 200,
    "name": "",
  },
}
`;

exports[`Optional & Partial Validation Object Validator Should validate null property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "name": "Mimi",
  },
}
`;

exports[`Optional & Partial Validation Object Validator Should validate undefined property 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "id",
    },
  ],
  "value": AnimalClass {
    "birthday": 2018-02-02,
    "deceased": true,
    "name": "Mimi",
  },
}
`;

exports[`Optional & Partial Validation Partial Validation Should pass empty field if parent marked with partial 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "id": 123,
  },
}
`;

exports[`Optional & Partial Validation Partial Validation Should skip optional validator on partial parent 1`] = `
Object {
  "issues": undefined,
  "value": AnimalClass {
    "id": 123,
  },
}
`;

exports[`Optional & Partial Validation Partial Validation Should validate whole object even its partial and provided undefined 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": undefined,
}
`;

exports[`Optional & Partial Validation Primitive Type Validation Should give proper info if provided wrong type of value 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Unable to convert \\"hula\\" into Number",
      ],
      "path": "data",
    },
  ],
  "value": "hula",
}
`;

exports[`Optional & Partial Validation Primitive Type Validation Should valid if optional 1`] = `
Object {
  "value": null,
}
`;

exports[`Optional & Partial Validation Primitive Type Validation Should validate null 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": null,
}
`;

exports[`Optional & Partial Validation Primitive Type Validation Should validate undefined 1`] = `
Object {
  "issues": Array [
    Object {
      "messages": Array [
        "Required",
      ],
      "path": "data",
    },
  ],
  "value": undefined,
}
`;
